File: arrays/4_nodo.js
// #region Listas_Enlazadas_JS
class Nodo {
    constructor(dato) {
        this.dato = dato;
        this.siguiente = null;
    }
}

class ListaEnlazada {
    constructor() {
        this.cabeza = null;
    }

    // Insertar al final (O(n) si no tenemos puntero al 'cola')
    insertarAlFinal(dato) {
        const nuevoNodo = new Nodo(dato);
        if (!this.cabeza) {
            this.cabeza = nuevoNodo;
            return;
        }
        let actual = this.cabeza;
        while (actual.siguiente) {
            actual = actual.siguiente;
        }
        actual.siguiente = nuevoNodo;
    }

    imprimir() {
        let actual = this.cabeza;
        let salida = "";
        while (actual) {
            salida += `[${actual.dato}] -> `;
            actual = actual.siguiente;
        }
        console.log(salida + "null");
    }
}

const miLista = new ListaEnlazada();
miLista.insertarAlFinal("Dato 1");
miLista.insertarAlFinal("Dato 2");
miLista.imprimir(); // [Dato 1] -> [Dato 2] -> null
// #endregion

File: avance/arrays/4_nodo.py
# #region Listas_Enlazadas_Python
class Nodo:
    def __init__(self, dato):
        self.dato = dato
        self.siguiente = None # El puntero al vecino

class ListaEnlazada:
    def __init__(self):
        self.cabeza = None # El inicio de la lista

    def insertar_al_inicio(self, dato):
        # O(1)
        nuevo_nodo = Nodo(dato)
        nuevo_nodo.siguiente = self.cabeza
        self.cabeza = nuevo_nodo

    def recorrer(self):
        # O(n)
        actual = self.cabeza
        while actual:
            print(f"[{actual.dato}] -> ", end="")
            actual = actual.siguiente
        print("None")

# Uso
lista = ListaEnlazada()
lista.insertar_al_inicio("Nodo C")
lista.insertar_al_inicio("Nodo B")
lista.insertar_al_inicio("Nodo A")
lista.recorrer() # [Nodo A] -> [Nodo B] -> [Nodo C] -> None
# #endregion